### ISSUE TITLE
Impure method is called for readonly field of value type

### ISSUE THREAD

**Issue by tvsonar** [SonarSource-VisualStudio/sonarlint-roslyn-analyzer#692]
R# _ImpureMethodCallOnReadonlyValueField_

Methods that don't change the visible state of an object and whose return value only depends on their input can be marked with the [Pure] attribute. If a method not marked with this attribute is called on a `readonly` field, which is a value type (so basically on a field that can't be changed), then the program might not behave as expected.

Example from: http://stackoverflow.com/questions/9927434/impure-method-is-called-for-readonly-field
```C#
public struct Nasty
{
    public int value;

    public void SetValue()
    {
        value = 10;
    }
}

class Test
{
    static readonly Nasty first;
    static Nasty second;

    static void Main()
    {
        first.SetValue();  // Noncompliant
        second.SetValue();
        Console.WriteLine(first.value);  // 0
        Console.WriteLine(second.value); // 10
    }
}
```

**Comment by ganncamp** [comment]
"impure" == does not have [Pure]

**Comment by ganncamp** [comment]
method of a value type that is marked `readonly` is called. if this method is `[Pure]`-able, then just add the attr. If not, the method won't actually work right (but no errors will be raised) and your code won't behave as expected

**Comment by ganncamp** [comment]
https://jira.sonarsource.com/browse/RSPEC-3465
